In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
order_items = pd.read_csv("/content/olist_order_items_dataset.csv")
orders = pd.read_csv("/content/olist_orders_dataset.csv")
products = pd.read_csv("/content/olist_products_dataset.csv")
customers = pd.read_csv("/content/olist_customers_dataset.csv")
geolocation = pd.read_csv("/content/olist_geolocation_dataset.csv")
order_payments = pd.read_csv("/content/olist_order_payments_dataset.csv")
order_reviews = pd.read_csv("/content/olist_order_reviews_dataset.csv")
sellers = pd.read_csv("/content/olist_sellers_dataset.csv")
product_category_name_translation = pd.read_csv("/content/product_category_name_translation.csv")


In [ ]:
df = pd.DataFrame(columns=['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm', 'product_category_name_english', 'payment_value', 'payment_type', 'review_score', 'delivery_time'])

In [ ]:
df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english,payment_value,payment_type,review_score,delivery_time


In [ ]:
merged_df = orders.merge(customers, on='customer_id', how='left')
merged_df = merged_df.merge(order_items, on='order_id', how='left')
merged_df = merged_df.merge(products, on='product_id', how='left')
merged_df = merged_df.merge(product_category_name_translation, on='product_category_name', how='left')
merged_df = merged_df.merge(sellers, on='seller_id', how='left')
merged_df = merged_df.merge(order_payments, on='order_id', how='left')
merged_df = merged_df.merge(order_reviews, on='order_id', how='left')

Now, let's look at the first few rows of the merged DataFrame and its shape to ensure the merge was successful.

In [ ]:
print(f"Shape of the merged DataFrame: {merged_df.shape}")
display(merged_df.head())

Shape of the merged DataFrame: (119143, 40)


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,payment_sequential,payment_type,payment_installments,payment_value,review_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,1.0,credit_card,1.0,18.12,a54f0611adc9ed256b57ede6b6eb5114,4.0,NaN,"Não testei o produto ainda, mas ele veio corre...",2017-10-11 00:00:00,2017-10-12 03:43:48
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,3.0,voucher,1.0,2.00,a54f0611adc9ed256b57ede6b6eb5114,4.0,NaN,"Não testei o produto ainda, mas ele veio corre...",2017-10-11 00:00:00,2017-10-12 03:43:48
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,2.0,voucher,1.0,18.59,a54f0611adc9ed256b57ede6b6eb5114,4.0,NaN,"Não testei o produto ainda, mas ele veio corre...",2017-10-11 00:00:00,2017-10-12 03:43:48
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,af07308b275d755c9edb36a90c618231,47813,...,1.0,boleto,1.0,141.46,8d5266042046a06655c8db133d120ba5,4.0,Muito boa a loja,Muito bom o produto.,2018-08-08 00:00:00,2018-08-08 18:37:50
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,1.0,credit_card,3.0,179.12,e73b67b67587f7644d5bd1a52deb1b01,5.0,NaN,NaN,2018-08-18 00:00:00,2018-08-22 19:07:58


In [ ]:
display(merged_df.head())
merged_df.info()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,payment_sequential,payment_type,payment_installments,payment_value,review_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,1.0,credit_card,1.0,18.12,a54f0611adc9ed256b57ede6b6eb5114,4.0,NaN,"Não testei o produto ainda, mas ele veio corre...",2017-10-11 00:00:00,2017-10-12 03:43:48
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,3.0,voucher,1.0,2.00,a54f0611adc9ed256b57ede6b6eb5114,4.0,NaN,"Não testei o produto ainda, mas ele veio corre...",2017-10-11 00:00:00,2017-10-12 03:43:48
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,7c396fd4830fd04220f754e42b4e5bff,3149,...,2.0,voucher,1.0,18.59,a54f0611adc9ed256b57ede6b6eb5114,4.0,NaN,"Não testei o produto ainda, mas ele veio corre...",2017-10-11 00:00:00,2017-10-12 03:43:48
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,af07308b275d755c9edb36a90c618231,47813,...,1.0,boleto,1.0,141.46,8d5266042046a06655c8db133d120ba5,4.0,Muito boa a loja,Muito bom o produto.,2018-08-08 00:00:00,2018-08-08 18:37:50
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,1.0,credit_card,3.0,179.12,e73b67b67587f7644d5bd1a52deb1b01,5.0,NaN,NaN,2018-08-18 00:00:00,2018-08-22 19:07:58


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119143 entries, 0 to 119142
Data columns (total 40 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       119143 non-null  object 
 1   customer_id                    119143 non-null  object 
 2   order_status                   119143 non-null  object 
 3   order_purchase_timestamp       119143 non-null  object 
 4   order_approved_at              118966 non-null  object 
 5   order_delivered_carrier_date   117057 non-null  object 
 6   order_delivered_customer_date  115722 non-null  object 
 7   order_estimated_delivery_date  119143 non-null  object 
 8   customer_unique_id             119143 non-null  object 
 9   customer_zip_code_prefix       119143 non-null  int64  
 10  customer_city                  119143 non-null  object 
 11  customer_state                 119143 non-null  object 
 12  order_item_id                 

In [ ]:
merged_df.isnull().count()

,0
order_id,119143
customer_id,119143
order_status,119143
order_purchase_timestamp,119143
order_approved_at,119143
order_delivered_carrier_date,119143
order_delivered_customer_date,119143
order_estimated_delivery_date,119143
customer_unique_id,119143
customer_zip_code_prefix,119143


In [ ]:
print("Number of missing values per column:")
print(merged_df.isnull().sum())
print("\nNumber of duplicate rows:")
print(merged_df.duplicated().sum())
print(merged_df.shape)
print(merged_df.describe())

Number of missing values per column:
order_id                              0
customer_id                           0
order_status                          0
order_purchase_timestamp              0
order_approved_at                   177
order_delivered_carrier_date       2086
order_delivered_customer_date      3421
order_estimated_delivery_date         0
customer_unique_id                    0
customer_zip_code_prefix              0
customer_city                         0
customer_state                        0
order_item_id                       833
product_id                          833
seller_id                           833
shipping_limit_date                 833
price                               833
freight_value                       833
product_category_name              2542
product_name_lenght                2542
product_description_lenght         2542
product_photos_qty                 2542
product_weight_g                    853
product_length_cm                   853
pro

Correct Cleaning

In [ ]:
merged_df = merged_df.drop_duplicates(
    subset=['order_id', 'order_item_id']
)

Convert Date Columns

In [ ]:
date_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date',
    'shipping_limit_date',
    'review_creation_date',
    'review_answer_timestamp'
]

for col in date_cols:
    merged_df[col] = pd.to_datetime(merged_df[col], errors='coerce')

#Handle Missing Values:
###Critical Columns
*   order_approved_at (177 missin)
    

   



In [ ]:
merged_df['order_approved_at'].fillna(
    merged_df['order_purchase_timestamp'], inplace=True
)

/tmp/ipykernel_13234/3258351006.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['order_approved_at'].fillna(


delivery dates

In [ ]:
merged_df['order_delivered_customer_date'].fillna(
    merged_df['order_estimated_delivery_date'], inplace=True
)

/tmp/ipykernel_13234/2942871264.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['order_delivered_customer_date'].fillna(


### Product-related (2542 missing)

In [ ]:
merged_df['product_category_name_english'].fillna("Unknown", inplace=True)

/tmp/ipykernel_13234/2095140216.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['product_category_name_english'].fillna("Unknown", inplace=True)


### Numeric Columns

In [ ]:
merged_df['review_score'].fillna(
    merged_df['review_score'].median(), inplace=True
)

merged_df['payment_value'].fillna(
    merged_df['payment_value'].median(), inplace=True
)

/tmp/ipykernel_13234/2310233911.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['review_score'].fillna(
/tmp/ipykernel_13234/2310233911.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.met

###Review Text Columns

In [ ]:
merged_df.drop(
    ['review_comment_title', 'review_comment_message'],
    axis=1,
    inplace=True
)

###Seller / Item Missing

In [ ]:
merged_df = merged_df.dropna(subset=[
    'product_id', 'seller_id', 'price'
])

###Remove Invalid Data

In [ ]:
# Remove zero/negative values
merged_df = merged_df[merged_df['price'] > 0]
merged_df = merged_df[merged_df['payment_value'] > 0]
merged_df = merged_df[merged_df['freight_value'] >= 0]

###Fix Product Dimensions

In [ ]:
merged_df = merged_df[
    (merged_df['product_weight_g'] > 0) &
    (merged_df['product_length_cm'] > 0) &
    (merged_df['product_height_cm'] > 0) &
    (merged_df['product_width_cm'] > 0)
]

###Outlier Removal

In [ ]:
Q1 = merged_df['payment_value'].quantile(0.25)
Q3 = merged_df['payment_value'].quantile(0.75)
IQR = Q3 - Q1

merged_df = merged_df[
    (merged_df['payment_value'] >= Q1 - 1.5 * IQR) &
    (merged_df['payment_value'] <= Q3 + 1.5 * IQR)
]

###Final Clean Columns

In [ ]:
merged_df.reset_index(drop=True, inplace=True)

In [ ]:
merged_df.describe()

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_zip_code_prefix,order_item_id,shipping_limit_date,price,freight_value,...,product_length_cm,product_height_cm,product_width_cm,seller_zip_code_prefix,payment_sequential,payment_installments,payment_value,review_score,review_creation_date,review_answer_timestamp
count,103454,103454,102447,103454,103454,103454.000000,103454.000000,103454,103454.000000,103454.000000,...,103454.000000,103454.000000,103454.000000,103454.000000,103451.000000,103451.000000,103454.000000,103454.000000,102617,102617
mean,2017-12-31 22:52:47.193022976,2018-01-01 09:10:52.775996928,2018-01-04 20:45:27.630374656,2018-01-13 13:15:28.630367488,2018-01-24 14:00:26.585729024,34731.651410,1.149448,2018-01-07 12:50:39.843824384,88.587297,18.410430,...,29.515959,15.756104,22.553579,23862.542463,1.024060,2.789021,123.783010,4.069586,2018-01-13 11:10:27.825798912,2018-01-16 14:20:08.748472320
min,2016-09-04 21:15:19,2016-09-15 12:16:38,2016-10-08 10:34:01,2016-10-11 13:46:32,2016-10-04 00:00:00,1003.000000,1.000000,2016-09-19 00:15:34,0.850000,0.000000,...,7.000000,2.000000,6.000000,1001.000000,1.000000,0.000000,0.010000,1.000000,2016-10-06 00:00:00,2016-10-07 18:32:28
25%,2017-09-13 16:05:05.249999872,2017-09-13 22:45:38.249999872,2017-09-18 18:02:41,2017-09-25 18:42:09,2017-10-04 00:00:00,11045.000000,1.000000,2017-09-20 15:10:59.750000128,38.400000,12.800000,...,18.000000,8.000000,15.000000,6097.000000,1.000000,1.000000,61.050000,4.000000,2017-09-26 00:00:00,2017-09-28 10:53:21
50%,2018-01-19 17:08:42,2018-01-20 09:09:26,2018-01-24 14:22:41,2018-02-01 22:51:34,2018-02-15 00:00:00,24110.000000,1.000000,2018-01-26 13:50:31.500000,69.000000,16.030000,...,24.000000,12.000000,20.000000,13483.000000,1.000000,2.000000,103.140000,5.000000,2018-02-02 00:00:00,2018-02-05 14:33:05
75%,2018-05-04 13:07:46,2018-05-04 19:34:22.500000,2018-05-08 11:31:00,2018-05-15 14:57:32.500000,2018-05-25 00:00:00,58025.000000,1.000000,2018-05-10 11:50:28.750000128,119.900000,19.800000,...,37.000000,20.000000,30.000000,25645.000000,1.000000,4.000000,166.670000,5.000000,2018-05-16 00:00:00,2018-05-20 12:06:38
max,2018-09-03 09:06:57,2018-09-03 17:40:06,2018-09-11 19:48:28,2018-10-17 13:22:46,2018-10-23 00:00:00,99990.000000,21.000000,2020-04-09 22:35:08,3899.000000,232.690000,...,105.000000,105.000000,118.000000,99730.000000,27.000000,24.000000,387.500000,5.000000,2018-08-31 00:00:00,2018-10-29 12:27:35
std,NaN,NaN,NaN,NaN,NaN,29810.447191,0.532243,NaN,71.773240,11.301125,...,15.626268,12.469287,11.253104,27161.588180,0.265698,2.572749,81.629263,1.361986,NaN,NaN


In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103454 entries, 0 to 103453
Data columns (total 38 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       103454 non-null  object        
 1   customer_id                    103454 non-null  object        
 2   order_status                   103454 non-null  object        
 3   order_purchase_timestamp       103454 non-null  datetime64[ns]
 4   order_approved_at              103454 non-null  datetime64[ns]
 5   order_delivered_carrier_date   102447 non-null  datetime64[ns]
 6   order_delivered_customer_date  103454 non-null  datetime64[ns]
 7   order_estimated_delivery_date  103454 non-null  datetime64[ns]
 8   customer_unique_id             103454 non-null  object        
 9   customer_zip_code_prefix       103454 non-null  int64         
 10  customer_city                  103454 non-null  object        
 11  

In [ ]:
merged_df.shape

(103454, 38)

###Fix Data Types

In [ ]:
cols_to_int = [
    'order_item_id',
    'payment_sequential',
    'payment_installments',
    'seller_zip_code_prefix'
]

for col in cols_to_int:
    if merged_df[col].isnull().any():
        merged_df[col].fillna(merged_df[col].median(), inplace=True)
    merged_df[col] = merged_df[col].astype('int64')

/tmp/ipykernel_13234/2199800069.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df[col].fillna(merged_df[col].median(), inplace=True)
/tmp/ipykernel_13234/2199800069.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inpl

###Handle Remaining Nulls

In [ ]:
merged_df.drop('product_category_name', axis=1, inplace=True)

In [ ]:
cols = [
    'product_name_lenght',
    'product_description_lenght',
    'product_photos_qty'
]

for col in cols:
    merged_df[col].fillna(merged_df[col].median(), inplace=True)

/tmp/ipykernel_13234/3131944460.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df[col].fillna(merged_df[col].median(), inplace=True)


review_id, review dates (~800 missing)

In [ ]:
merged_df['review_creation_date'].fillna(
    merged_df['order_delivered_customer_date'], inplace=True
)

merged_df['review_answer_timestamp'].fillna(
    merged_df['review_creation_date'], inplace=True
)

/tmp/ipykernel_13234/3642017393.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['review_creation_date'].fillna(
/tmp/ipykernel_13234/3642017393.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using

Standardize Categorical Data

In [ ]:
merged_df['order_status'] = merged_df['order_status'].str.lower().str.strip()
merged_df['payment_type'] = merged_df['payment_type'].str.lower().str.strip()

Remove Irrelevant Columns

In [ ]:
cols_to_drop = [
    'review_id',
    'customer_zip_code_prefix',
    'seller_zip_code_prefix'
]

merged_df.drop(cols_to_drop, axis=1, inplace=True)

In [ ]:
merged_df['order_delivered_carrier_date'].fillna(
    merged_df['order_approved_at'], inplace=True
)

/tmp/ipykernel_13234/3618685295.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['order_delivered_carrier_date'].fillna(


In [ ]:
merged_df['payment_type'].fillna("unknown", inplace=True)

/tmp/ipykernel_13234/404402676.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_df['payment_type'].fillna("unknown", inplace=True)


In [ ]:
cat_cols = [
    'order_status',
    'customer_city',
    'customer_state',
    'product_category_name_english',
    'seller_city',
    'seller_state',
    'payment_type'
]

for col in cat_cols:
    merged_df[col] = merged_df[col].astype('category')

In [ ]:
cols_to_drop = [
    'customer_id',
    'seller_id'
]

merged_df.drop(cols_to_drop, axis=1, inplace=True)

Final Sanity Checks

In [ ]:
print("Final Shape:", merged_df.shape)

print("\nNull values:")
print(merged_df.isnull().sum())

print("\nData types:")
print(merged_df.dtypes)

Final Shape: (103454, 32)

Null values:
order_id                         0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
customer_unique_id               0
customer_city                    0
customer_state                   0
order_item_id                    0
product_id                       0
shipping_limit_date              0
price                            0
freight_value                    0
product_name_lenght              0
product_description_lenght       0
product_photos_qty               0
product_weight_g                 0
product_length_cm                0
product_height_cm                0
product_width_cm                 0
product_category_name_english    0
seller_city                      0
seller_state                     0
payment_sequential               0
payment_type                     0
payment_install

#**# FEATURE ENGINEERING**

Order-Level Features

In [ ]:
df = merged_df.copy()

# Total order value
df['total_value'] = df['price'] + df['freight_value']

# Delivery time (days)
df['delivery_time'] = (
    df['order_delivered_customer_date'] - df['order_purchase_timestamp']
).dt.days

# Approval time (processing delay)
df['approval_time'] = (
    df['order_approved_at'] - df['order_purchase_timestamp']
).dt.seconds / 3600  # hours

Delivery Performance Features

In [ ]:
# Delivery delay (late or early)
df['delivery_delay'] = (
    df['order_delivered_customer_date'] - df['order_estimated_delivery_date']
).dt.days

# Late delivery flag
df['is_late'] = (df['delivery_delay'] > 0).astype(int)

Time-Based Features

In [ ]:
df['order_year'] = df['order_purchase_timestamp'].dt.year
df['order_month'] = df['order_purchase_timestamp'].dt.month
df['order_day'] = df['order_purchase_timestamp'].dt.day
df['order_weekday'] = df['order_purchase_timestamp'].dt.weekday
df['order_hour'] = df['order_purchase_timestamp'].dt.hour

Customer-Level Aggregation (RFM BASE 🔥)

In [ ]:
customer_df = df.groupby('customer_unique_id').agg({
    'order_purchase_timestamp': ['max', 'count'],
    'payment_value': 'sum',
    'review_score': 'mean'
}).reset_index()

customer_df.columns = [
    'customer_id',
    'last_purchase',
    'frequency',
    'monetary',
    'avg_review'
]

Calculate Recency

In [ ]:
import datetime

latest_date = df['order_purchase_timestamp'].max()

customer_df['recency'] = (
    latest_date - customer_df['last_purchase']
).dt.days

RFM Segmentation

In [ ]:
# Recency score
r_cut = pd.qcut(customer_df['recency'], 4, duplicates='drop')
r_labels = list(range(len(r_cut.cat.categories), 0, -1)) if len(r_cut.cat.categories) > 0 else []
customer_df['R_score'] = pd.qcut(customer_df['recency'], 4, labels=r_labels, duplicates='drop')

# Frequency score
f_cut = pd.qcut(customer_df['frequency'], 4, duplicates='drop')
f_labels = list(range(1, len(f_cut.cat.categories) + 1)) if len(f_cut.cat.categories) > 0 else []
customer_df['F_score'] = pd.qcut(customer_df['frequency'], 4, labels=f_labels, duplicates='drop')

# Monetary score
m_cut = pd.qcut(customer_df['monetary'], 4, duplicates='drop')
m_labels = list(range(1, len(m_cut.cat.categories) + 1)) if len(m_cut.cat.categories) > 0 else []
customer_df['M_score'] = pd.qcut(customer_df['monetary'], 4, labels=m_labels, duplicates='drop')

customer_df['RFM_score'] = (
    customer_df['R_score'].astype(str) +
    customer_df['F_score'].astype(str) +
    customer_df['M_score'].astype(str)
)

In [ ]:
customer_df.head()

,customer_id,last_purchase,frequency,monetary,avg_review,recency,R_score,F_score,M_score,RFM_score
0,0000366f3b9a7992bf8c76cfdf3221e2,2018-05-10 10:56:27,1,141.90,5.0,115,4,1,3,413
1,0000b849f77a49e4a4ce2b2a4ca5be3f,2018-05-07 11:11:27,1,27.19,4.0,118,4,1,1,411
2,0000f46a3911fa3c0805444483337064,2017-03-10 21:05:03,1,86.22,3.0,541,1,1,2,112
3,0000f6ccb0745a6a4b88665a16c9f078,2017-10-12 20:29:41,1,43.62,4.0,325,2,1,1,211
4,0004aac84e0df4da2b147fca70cf8255,2017-11-14 19:45:42,1,196.89,5.0,292,2,1,4,214


Product-Level Features (Inventory + Demand)

In [ ]:
product_df = df.groupby('product_id').agg({
    'order_id': 'count',
    'payment_value': 'sum',
    'price': 'mean'
}).reset_index()

product_df.columns = [
    'product_id',
    'total_orders',
    'total_revenue',
    'avg_price'
]

Daily Sales (FOR FORECASTING)

In [ ]:
time_series_df = df.groupby(
    df['order_purchase_timestamp'].dt.date
).agg({
    'payment_value': 'sum'
}).reset_index()

time_series_df.columns = ['date', 'daily_sales']

In [ ]:
df.head()

,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_city,customer_state,...,total_value,delivery_time,approval_time,delivery_delay,is_late,order_year,order_month,order_day,order_weekday,order_hour
0,e481f51cbdc54678b7cc49136f2d6af7,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,sao paulo,SP,...,38.71,8,0.178333,-8,0,2017,10,2,0,10
1,53cdb2fc8bc7dce0b6741e2150273451,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,barreiras,BA,...,141.46,13,6.713889,-6,0,2018,7,24,1,20
2,47770eb9100c2d0c44946d9cf07ec65d,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,vianopolis,GO,...,179.12,9,0.276111,-18,0,2018,8,8,2,8
3,949d5b44dbf5de918fe9c16f97b45f8a,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,sao goncalo do amarante,RN,...,72.20,13,0.298056,-13,0,2017,11,18,5,19
4,ad21c59c0840e6cb83a9ceb5573f8159,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,santo andre,SP,...,28.62,2,1.030556,-10,0,2018,2,13,1,21


In [ ]:
df.shape

(103454, 42)

In [ ]:
time_series_df.head()

,date,daily_sales
0,2016-09-04,272.46
1,2016-09-05,75.06
2,2016-09-15,337.32
3,2016-10-02,109.34
4,2016-10-03,595.14


In [ ]:
product_df.head(

)

,product_id,total_orders,total_revenue,avg_price
0,00066f42aeeb9f3007548bb9d3f33c38,1,120.24,101.65
1,00088930e925c41fd95ebfe695fd2655,1,143.83,129.90
2,0009406fd7479715e4bef61dd91f2462,1,242.10,229.00
3,000b8f95fcb9e0096488278317764d19,2,157.00,58.90
4,000d9be29b5207b54e86aa1b1ac54872,1,218.27,199.00


In [ ]:
df.to_csv("orders.csv", index=False)
customer_df.to_csv("customers.csv", index=False)
product_df.to_csv("products.csv", index=False)
time_series_df.to_csv("time_series.csv", index=False)

In [ ]:
import zipfile

with zipfile.ZipFile('RetailPulse_Data.zip', 'w') as z:
    z.write("orders.csv")
    z.write("customers.csv")
    z.write("products.csv")
    z.write("time_series.csv")

In [ ]:
from google.colab import files
files.download("RetailPulse_Data.zip")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>